src/index.js

```diff
@@ -609,12 +610,40 @@ const members = {
       const getHtmlNodeOuterHtml = (node) => node.outerHTML;
       const getOuterHtmlByNodeType = {
         1: getHtmlNodeOuterHtml,
-        3: getEscapedTextContent,
         8: getCommentNodeOuterHtml,
       };
+      // Text nodes with these ancestors should be treated as raw text
+      // See section 8.4 of
+      // https://www.w3.org/TR/2008/WD-html5-20080610/serializing.html#html-fragment
+      const rawTextNodeAncestors = {
+        style: true,
+        script: true,
+        xmp: true,
+        iframe: true,
+        noembed: true,
+        noframes: true,
+        noscript: true,
+        plaintext: true,
+      };
 
       eachChildNode(this, node => {
-        const getOuterHtml = getOuterHtmlByNodeType[node.nodeType] || getHtmlNodeOuterHtml;
+        let getOuterHtml;
+        if (node.nodeType === 3) {
+          let parentNode = node.parentNode;
+          while (parentNode) {
+            if (parentNode.nodeType === 1 && // only for elements
+                parentNode.nodeName.toLowerCase() in rawTextNodeAncestors) {
+              getOuterHtml = getRawTextContent;
```

Comment:
For all of those cases can't you cache whether or not the current node is one of those or is a descendant of one of those types and then do stuff with the children based on that value instead of checking within the loop?